contract/src/RadicleCloudBase.sol

```diff
@@ -0,0 +1,264 @@
+    ) internal pure returns (uint64) {
+        return (oldRate * oldLen + newRate * newLen) / (oldLen + newLen);
+    }
+
+    /// @notice Renew expiry or create a new deployment.
+    /// @param org the address of org we're topping up
+    /// @param _owner the address of org owner (who gets refunds)
+    /// @param amount the amount being topped up (must be exact multiples of ratePerBlock * duration)
+    function topUp(
+        address org,
+        address _owner,
+        uint128 amount
+    ) internal {
+        if (amount >= ratePerBlock * duration) {
+            if (dep[org].expiry > block.number) {
+                // active user
+                for (uint64 i = 0; i < amount / (ratePerBlock * duration); ++i) {
+                    hasRenewal[dep[org].expiry + (i * duration)] = true;
+                }
+
+                uint64 oldRate = dep[org].rate;
+                if (ratePerBlock != oldRate) {
+                    dep[org].rate = weightedRate(
+                        oldRate,
+                        ratePerBlock,
+                        uint64(dep[org].expiry - block.number),
+                        duration
+                    );
+                    // create new impact point in renews
+                    renews.add(uint64(block.number));
+                    // set its value to rate delta
+                    impacts[uint64(block.number)] += int64(dep[org].rate) - int64(oldRate);
+                }
+
+                // cancel out old expiry
+                impacts[dep[org].expiry] += int64(oldRate);
+                dep[org].expiry += duration;
+                // create impact of new expiry
+                impacts[dep[org].expiry] -= int64(dep[org].rate);
+            } else {
+                // new user
+                dep[org].expiry = uint64(block.number + duration);
+                dep[org].start = uint64(block.number);
+                dep[org].owner = _owner;
+                dep[org].rate = ratePerBlock;
+
+                buys.add(uint64(block.number));
+                expires.add((uint64(block.number + duration)));
+
+                impacts[uint64(block.number)] += int64(ratePerBlock);
+                impacts[uint64(block.number + duration)] -= int64(ratePerBlock);
+            }
+            emit NewTopUp(org, dep[org].expiry);
+        }
+    }
+
+    /// @dev Smaller between `a` and `b`.
+    function min(uint64 a, uint64 b) internal pure returns (uint64) {
+        return a < b ? a : b;
+    }
+
+    /// @dev Return next impact point for rate.
+    /// @return smallest block between next buy, expiry, cancel, and renew
+    function _nextImpact() internal view returns (uint64) {
+        return min(min(buys.last(), expires.last()), min(cancels.last(), renews.last()));
+    }
+
+    /// @dev Calculate the amount operator has earned till current block.
+    /// @return revenue earned from last processed blocked till now
+    function _calcRealizedRevenue() internal returns (uint128) {
```

Comment:
We're using both uint128s and uint64 for holding balances/amounts, I'm not sure why.. It would be easier to hold `uint256` everywhere, since that is the type used for value and balance in ERC20 contracts.